### PR TITLE
Storage rocksdb

### PR DESCRIPTION
**Pull request by tsvisabo**
Basic implementation of storage trait

TODO: 
- [x] Storage: add methods for fetching transaction snapshot_index and solid state, implement for both backends
- [x] Add tests for state_delta, solid state and snapshot index update
- [x] Add column families for trunk/branch (rocksdb)?
- [x] cleanup

### CODE REVIEW

File: bee-storage/src/rocksdb/mod.rs [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+            cast_slice_mut(hash.0.as_mut()).as_mut_ptr(),
+            HASH_TRIT_LEN,
+        )
+    };
+
+    hash
+}
+
+#[inline]
+fn encode_transaction(tx: &Transaction, mut buf: &mut Trits<T1B1>) {
```
Comment (thibault-martinez) [inline bee-storage/src/rocksdb/mod.rs:149]: Is this basically `to_trits` ? Should probably be in the bundle crate since we are still missing it there.

Comment (tsvisabo) [reply bee-storage/src/rocksdb/mod.rs:149]: It is, and i think it should indeed be in bundle's,
i just don't like it's asymmetric with from_trits, this one takes an already allocated buffer

Comment (thibault-martinez) [reply bee-storage/src/rocksdb/mod.rs:149]: I think that's usually the difference in naming them `to_*` and `into_*`, we can provide both.

File: bee-bundle/src/constants.rs [hunk 2]
```diff
@@ -67,21 +67,21 @@ pub const INDEX_TRIT_LEN: usize = 27;
...
-pub(crate) const LAST_INDEX: Field = offsets_from_previous_field!(INDEX, INDEX_TRIT_LEN);
-pub(crate) const BUNDLE: Field = offsets_from_previous_field!(LAST_INDEX, HASH_TRIT_LEN);
-pub(crate) const TRUNK: Field = offsets_from_previous_field!(BUNDLE, HASH_TRIT_LEN);
-pub(crate) const BRANCH: Field = offsets_from_previous_field!(TRUNK, HASH_TRIT_LEN);
-pub(crate) const TAG: Field = offsets_from_previous_field!(BRANCH, TAG_TRIT_LEN);
-pub(crate) const ATTACHMENT_TS: Field = offsets_from_previous_field!(TAG, TIMESTAMP_TRIT_LEN);
-pub(crate) const ATTACHMENT_LBTS: Field = offsets_from_previous_field!(ATTACHMENT_TS, TIMESTAMP_TRIT_LEN);
-pub(crate) const ATTACHMENT_UBTS: Field = offsets_from_previous_field!(ATTACHMENT_LBTS, TIMESTAMP_TRIT_LEN);
-pub(crate) const NONCE: Field = offsets_from_previous_field!(ATTACHMENT_UBTS, NONCE_TRIT_LEN);
+pub const PAYLOAD: Field = offsets_from_trits!(0, PAYLOAD_TRIT_LEN);
```
Comment (thibault-martinez) [inline bee-bundle/src/constants.rs:70]: Is this needed ?

Comment (tsvisabo) [reply bee-bundle/src/constants.rs:70]: not at all, will undo

File: bee-storage/src/rocksdb/mod.rs [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    }
+}
+
+#[async_trait]
+impl Connection<RocksDBBackendConnection> for RocksDBBackendConnection {
+    type StorageError = RocksDbBackendError;
+
+    async fn establish_connection(&mut self, url: &str) -> Result<(), RocksDbBackendError> {
+        let mut cf_opts = Options::default();
+        cf_opts.set_max_write_buffer_number(16);
```
[the remark this answers is not in the document]
Comment (tsvisabo) [reply bee-storage/src/rocksdb/mod.rs:109]: yeah, missed that

File: bee-storage/src/rocksdb/mod.rs [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+            &milestone_cf_hash_to_index,
+            cast_slice(milestone_hash.to_inner().encode::<T5B1Buf>().as_i8_slice()),
+        )?;
+
+        if res.is_none() {
+            return Err(RocksDbBackendError::TransactionDoesNotExist);
+        }
+
+        let mut index_buf: [u8; 4] = [0; 4];
+        unsafe { ptr::copy(res.unwrap().as_slice().as_ptr(), index_buf.as_mut_ptr(), 4) };
```
Comment (wusyong) [inline bee-storage/src/rocksdb/mod.rs:495]: Probably safer to use `copy_from_slice` instead
```suggestion
        index_buf.copy_from_slice(res.unwrap().as_slice());
```

File: bee-storage/src/rocksdb/mod.rs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+    Options,
+    WriteOptions,
+    DB,
+};
+use std::borrow::BorrowMut;
+
+const TRANSACTION_CF_HASH_TO_TRANSACTION: &str = "transaction_hash_to_transaction";
+const TRANSACTION_CF_HASH_TO_SOLID: &str = "transaction_hash_to_solid";
+const TRANSACTION_CF_HASH_TO_SNAPSHOT_INDEX: &str = "transaction_hash_to_snapshot_index";
+const TRANSACTION_CF_HASH_TO_APROVEES: &str = "transaction_hash_to_trunk";
```
Comment (thibault-martinez) [inline bee-storage/src/rocksdb/mod.rs:66]: rename

Comment (tsvisabo) [reply bee-storage/src/rocksdb/mod.rs]: horrible..
thanks

**Pull request merged**